### PR TITLE
Fix summary reporting with parallel replicas with LIMIT

### PR DESCRIPTION
**Pull request by Algunenano**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Fix summary reporting with parallel replicas with LIMIT

### Documentation entry for user-facing changes

When using parallel replicas the summary might be reported incompletely:
```bash
echo "SELECT * FROM replicated_numbers limit 20 SETTINGS max_parallel_replicas = 2, cluster_for_parallel_replicas = 'test_cluster_one_shard_three_replicas_localhost', allow_experimental_parallel_reading_from_replicas = 1, use_hedged_requests = 0" | curl 'http://clickhouse-02:58123/?wait_end_of_query=1' --data-binary @- -vvv 2>&1 | grep Summary
< X-ClickHouse-Summary: {"read_rows":"0","read_bytes":"0","written_rows":"0","written_bytes":"0","total_rows_to_read":"0","result_rows":"20","result_bytes":"512","peak_memory_usage":"4227652"}
```

The problem is that once the coordinator has read enough rows from the replicas it cancels all the queries and discards any pending packets, so the progress isn't accounted. It then doesn't match the values in system.query_log for all the queries.

2 things needed to be changed:
* On cancel, make the coordinator read and process any pending progress packets instead of discard them via drain().
* On pipelineExecutor finish(), check if there is any progress that hasn't been processed and report it. This is needed to collect all pending reports generated asynchronously but reported when finishing the connection.

Note that with this new system it might end up reporting more rows than necessary because the read happens in parallel and each replica will read rows concurrently. Same might happen when doing local reads with concurrent reads (`merge_tree_min_rows_for_concurrent_read`), so it's the same behaviour now.

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
This is an automated comment for commit c8a9284abc3c2b084d983660688281efd04e597f with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/53050/c8a9284abc3c2b084d983660688281efd04e597f/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by Algunenano** [comment]
Seems there is a race condition when calling progress:

```
==================
WARNING: ThreadSanitizer: data race (pid=688)
  Write of size 1 at 0x7b4c011d57b0 by thread T674 (mutexes: write M0, write M1):
    #0 DB::ISource::progress(unsigned long, unsigned long) build_docker/./src/Processors/ISource.cpp:68:27 (clickhouse+0x1b299c2a) (BuildId: c3fd0d98af1a5e6225d18d6d5f45206cdd1c8f2d)
    #1 DB::RemoteSource::tryGenerate()::$_0::operator()(DB::Progress const&) const build_docker/./src/Processors/Sources/RemoteSource.cpp:82:13 (clickhouse+0x1b719345) (BuildId: c3fd0d98af1a5e6225d18d6d5f45206cdd1c8f2d)
    #2 decltype(std::declval<DB::RemoteSource::tryGenerate()::$_0&>()(std::declval<DB::Progress const&>())) std::__1::__invoke[abi:v15000]<DB::RemoteSource::tryGenerate()::$_0&, DB::Progress const&>(DB::RemoteSource::tryGenerate()::$_0&, DB::Progress const&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0x1b719345)
    #3 void std::__1::__invoke_void_return_wrapper<void, true>::__call<DB::RemoteSource::tryGenerate()::$_0&, DB::Progress const&>(DB::RemoteSource::tryGenerate()::$_0&, DB::Progress const&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479:9 (clickhouse+0x1b719345)
    #4 std::__1::__function::__default_alloc_func<DB::RemoteSource::tryGenerate()::$_0, void (DB::Progress const&)>::operator()[abi:v15000](DB::Progress const&) build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235:12 (clickhouse+0x1b719345)
    #5 void std::__1::__function::__policy_invoker<void (DB::Progress const&)>::__call_impl<std::__1::__function::__default_alloc_func<DB::RemoteSource::tryGenerate()::$_0, void (DB::Progress const&)>>(std::__1::__function::__policy_storage const*, DB::Progress const&) build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:716:16 (clickhouse+0x1b719345)
    #6 std::__1::__function::__policy_func<void (DB::Progress const&)>::operator()[abi:v15000](DB::Progress const&) const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:848:16 (clickhouse+0x186f9ad3) (BuildId: c3fd0d98af1a5e6225d18d6d5f45206cdd1c8f2d)
    #7 std::__1::function<void (DB::Progress const&)>::operator()(DB::Progress const&) const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:1187:12 (clickhouse+0x186f9ad3)
    #8 DB::RemoteQueryExecutor::finish() build_docker/./src/QueryPipeline/RemoteQueryExecutor.cpp:638:21 (clickhouse+0x186f9ad3)
    #9 DB::RemoteSource::onUpdatePorts() build_docker/./src/Processors/Sources/RemoteSource.cpp:181:25 (clickhouse+0x1b717b79) (BuildId: c3fd0d98af1a5e6225d18d6d5f45206cdd1c8f2d)
    #10 DB::ExecutingGraph::updateNode(unsigned long, std::__1::queue<DB::ExecutingGraph::Node*, std::__1::deque<DB::ExecutingGraph::Node*, std::__1::allocator<DB::ExecutingGraph::Node*>>>&, std::__1::queue<DB::ExecutingGraph::Node*, std::__1::deque<DB::ExecutingGraph::Node*, std::__1::allocator<DB::ExecutingGraph::Node*>>>&) build_docker/./src/Processors/Executors/ExecutingGraph.cpp:255:49 (clickhouse+0x1b2b5aac) (BuildId: c3fd0d98af1a5e6225d18d6d5f45206cdd1c8f2d)
    #11 DB::PipelineExecutor::executeStepImpl(unsigned long, std::__1::atomic<bool>*) build_docker/./src/Processors/Executors/PipelineExecutor.cpp:295:29 (clickhouse+0x1b2af6e5) (BuildId: c3fd0d98af1a5e6225d18d6d5f45206cdd1c8f2d)
    #12 DB::PipelineExecutor::executeSingleThread(unsigned long) build_docker/./src/Processors/Executors/PipelineExecutor.cpp:242:5 (clickhouse+0x1b2b09b0) (BuildId: c3fd0d98af1a5e6225d18d6d5f45206cdd1c8f2d)
    #13 DB::PipelineExecutor::spawnThreads()::$_0::operator()() const build_docker/./src/Processors/Executors/PipelineExecutor.cpp:364:17 (clickhouse+0x1b2b09b0)
    #14 decltype(std::declval<DB::PipelineExecutor::spawnThreads()::$_0&>()()) std::__1::__invoke[abi:v15000]<DB::PipelineExecutor::spawnThreads()::$_0&>(DB::PipelineExecutor::spawnThreads()::$_0&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0x1b2b09b0)
    #15 void std::__1::__invoke_void_return_wrapper<void, true>::__call<DB::PipelineExecutor::spawnThreads()::$_0&>(DB::PipelineExecutor::spawnThreads()::$_0&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479:9 (clickhouse+0x1b2b09b0)
    #16 std::__1::__function::__default_alloc_func<DB::PipelineExecutor::spawnThreads()::$_0, void ()>::operator()[abi:v15000]() build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235:12 (clickhouse+0x1b2b09b0)
    #17 void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<DB::PipelineExecutor::spawnThreads()::$_0, void ()>>(std::__1::__function::__policy_storage const*) build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:716:16 (clickhouse+0x1b2b09b0)
    #18 std::__1::__function::__policy_func<void ()>::operator()[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:848:16 (clickhouse+0x10119fd1) (BuildId: c3fd0d98af1a5e6225d18d6d5f45206cdd1c8f2d)
    #19 std::__1::function<void ()>::operator()() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:1187:12 (clickhouse+0x10119fd1)
    #20 ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::worker(std::__1::__list_iterator<ThreadFromGlobalPoolImpl<false>, void*>) build_docker/./src/Common/ThreadPool.cpp:427:13 (clickhouse+0x10119fd1)
    #21 void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()::operator()() const build_docker/./src/Common/ThreadPool.cpp:180:73 (clickhouse+0x1011edbc) (BuildId: c3fd0d98af1a5e6225d18d6d5f45206cdd1c8f2d)
    #22 decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()&>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0x1011edbc)
    #23 decltype(auto) std::__1::__apply_tuple_impl[abi:v15000]<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()&, std::__1::tuple<>&>(void&&, std::__1::tuple<>&, std::__1::__tuple_indices<>) build_docker/./contrib/llvm-project/libcxx/include/tuple:1789:1 (clickhouse+0x1011edbc)
    #24 decltype(auto) std::__1::apply[abi:v15000]<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()&, std::__1::tuple<>&>(void&&, std::__1::tuple<>&) build_docker/./contrib/llvm-project/libcxx/include/tuple:1798:1 (clickhouse+0x1011edbc)
    #25 ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()::operator()() build_docker/./src/Common/ThreadPool.h:242:13 (clickhouse+0x1011edbc)
    #26 decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()&>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0x1011ed21) (BuildId: c3fd0d98af1a5e6225d18d6d5f45206cdd1c8f2d)
    #27 void std::__1::__invoke_void_return_wrapper<void, true>::__call<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()&>(ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479:9 (clickhouse+0x1011ed21)
    #28 std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>::operator()[abi:v15000]() build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235:12 (clickhouse+0x1011ed21)
    #29 void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>>(std::__1::__function::__policy_storage const*) build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:716:16 (clickhouse+0x1011ed21)
    #30 std::__1::__function::__policy_func<void ()>::operator()[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:848:16 (clickhouse+0x10115cc6) (BuildId: c3fd0d98af1a5e6225d18d6d5f45206cdd1c8f2d)
    #31 std::__1::function<void ()>::operator()() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:1187:12 (clickhouse+0x10115cc6)
    #32 ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) build_docker/./src/Common/ThreadPool.cpp:427:13 (clickhouse+0x10115cc6)
    #33 void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()::operator()() const build_docker/./src/Common/ThreadPool.cpp:180:73 (clickhouse+0x1011be71) (BuildId: c3fd0d98af1a5e6225d18d6d5f45206cdd1c8f2d)
    #34 decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0x1011be71)
    #35 void std::__1::__thread_execute[abi:v15000]<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(std::__1::tuple<void, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>&, std::__1::__tuple_indices<>) build_docker/./contrib/llvm-project/libcxx/include/thread:284:5 (clickhouse+0x1011be71)
    #36 void* std::__1::__thread_proxy[abi:v15000]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>>(void*) build_docker/./contrib/llvm-project/libcxx/include/thread:295:5 (clickhouse+0x1011be71)

  Previous write of size 1 at 0x7b4c011d57b0 by thread T1007:
    #0 DB::ISource::work() build_docker/./src/Processors/ISource.cpp:92:31 (clickhouse+0x1b299e4b) (BuildId: c3fd0d98af1a5e6225d18d6d5f45206cdd1c8f2d)
    #1 DB::executeJob(DB::ExecutingGraph::Node*, DB::ReadProgressCallback*) build_docker/./src/Processors/Executors/ExecutionThreadContext.cpp:47:26 (clickhouse+0x1b2bcd74) (BuildId: c3fd0d98af1a5e6225d18d6d5f45206cdd1c8f2d)
    #2 DB::ExecutionThreadContext::executeTask() build_docker/./src/Processors/Executors/ExecutionThreadContext.cpp:95:9 (clickhouse+0x1b2bcd74)
    #3 DB::PipelineExecutor::executeStepImpl(unsigned long, std::__1::atomic<bool>*) build_docker/./src/Processors/Executors/PipelineExecutor.cpp:276:26 (clickhouse+0x1b2af9b0) (BuildId: c3fd0d98af1a5e6225d18d6d5f45206cdd1c8f2d)
    #4 DB::PipelineExecutor::executeSingleThread(unsigned long) build_docker/./src/Processors/Executors/PipelineExecutor.cpp:242:5 (clickhouse+0x1b2b09b0) (BuildId: c3fd0d98af1a5e6225d18d6d5f45206cdd1c8f2d)
    #5 DB::PipelineExecutor::spawnThreads()::$_0::operator()() const build_docker/./src/Processors/Executors/PipelineExecutor.cpp:364:17 (clickhouse+0x1b2b09b0)
    #6 decltype(std::declval<DB::PipelineExecutor::spawnThreads()::$_0&>()()) std::__1::__invoke[abi:v15000]<DB::PipelineExecutor::spawnThreads()::$_0&>(DB::PipelineExecutor::spawnThreads()::$_0&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0x1b2b09b0)
    #7 void std::__1::__invoke_void_return_wrapper<void, true>::__call<DB::PipelineExecutor::spawnThreads()::$_0&>(DB::PipelineExecutor::spawnThreads()::$_0&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479:9 (clickhouse+0x1b2b09b0)
    #8 std::__1::__function::__default_alloc_func<DB::PipelineExecutor::spawnThreads()::$_0, void ()>::operator()[abi:v15000]() build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235:12 (clickhouse+0x1b2b09b0)
--
==================
```

I'll fix it after I come back from holidays. Or maybe after

**Comment by Algunenano** [comment]
`Stateless tests (tsan) [4/5] — Tests are not finished, fail: 1, passed: 73` -> Unrelated bug detected with TSAN -> https://github.com/ClickHouse/ClickHouse/issues/53827

**Comment by Algunenano** [comment]
> Stateless tests (release, s3 storage) [2/2] -> Seems unrelated as it's stuck dropping databases. I've created https://github.com/ClickHouse/ClickHouse/issues/53894 to investigate further.

**Review by nikitamikhaylov — APPROVED** [review]

File: src/Processors/ISource.cpp [hunk 1]
```diff
@@ -85,6 +87,18 @@ std::optional<ISource::ReadProgress> ISource::getReadProgress()
     return ReadProgress{res_progress, empty_limits};
 }
 
+void ISource::addTotalRowsApprox(size_t value)
+{
+    std::lock_guard lock(read_progress_mutex);
```
Comment (devcrafter) [inline src/Processors/ISource.cpp:92]: @Algunenano Could you elaborate why the mutex is necessary? The `ISource` interface is not supposed to be accessed concurrently.

Comment (Algunenano) [reply src/Processors/ISource.cpp:92]: You can see a sample report in https://github.com/ClickHouse/ClickHouse/pull/53050#issuecomment-1671355082 were ISource::work and ISource::progress are being called concurrently.

Comment (devcrafter) [reply src/Processors/ISource.cpp:92]: Ok, but wasn't it introduced by this PR here? - https://github.com/ClickHouse/ClickHouse/pull/53050/files#diff-391fc60bfa23a1270bda630bb11853d9d50915a10b6bb66b2d116a3e1a287e02R209

Comment (Algunenano) [reply src/Processors/ISource.cpp:92]: I'm not sure about what's wrong with the approach and I'm AFK for a few weeks. If you think it's the wrong approach we can revert and revisit later

**Pull request merged**